packages/keystone/apolloErrorFormatter.js

```diff
@@ -96,130 +97,222 @@ function _patchKnownErrorCases (error, result) {
+    const pickKeys1 = (hideInternals) ? ['message', 'name'] : ['message', 'name', 'stack']
+    Object.assign(result, pick(error, pickKeys1))
+
+    // [2] base graphql fields: locations, path
+    //  - `locations` - array of { line, column } locations
+    //  - `path` - array describing the JSON-path
+    const pickKeys3 = ['locations', 'path']
+    Object.assign(result, pick(error, pickKeys3))
+
+    // [3] graphql extensions field
+    const hasErrorExtensions = error.extensions && !isEmpty(error.extensions)
+    const hasOriginalErrorExtensions = originalError && !isEmpty(originalError.extensions)
+    if (hasErrorExtensions || hasOriginalErrorExtensions) {
+        if (hasErrorExtensions) {
+            Object.assign(extensions, error.extensions)
+        }
+        if (hasOriginalErrorExtensions) {
+            Object.assign(extensions, originalError.extensions)
+        }
     }
 
+    // [4] apollo/keystone error fields: time_thrown, data, internalData (it's old ApolloServer keys)
+    const pickKeys2 = (hideInternals) ? ['data'] : ['data', 'time_thrown', 'internalData']
+    Object.assign(result, pick(error, pickKeys2))
 
+    // [5] add messageForDeveloper field: if has (nodes) or has (source and location) => can use printError
+    if (!_isRecursionCall && (error.nodes || (error.source && error.location))) {
+        const messageForDeveloper = printError(error)
+        if (messageForDeveloper !== error.message) {
+            extensions.messageForDeveloper = messageForDeveloper
+        }
+    }
 
-    // KeystoneJS hotfixes! Taken from KeystoneJS sources. Probably useless in a future but we already have a tests for that!
-    if (originalError) {
-        if (originalError.path && !result.path) {
-            result.path = originalError.path
+    // NOTE(pahaz): ApolloServer wraps all errors by GraphQLError. We want to change the results based on the wrapped error
+    // It's backward API compatibility fixes. Really you should always use `GQLError` but at the moment we have some old throw error code.
+    if (!_isRecursionCall && errorCName === 'GraphQLError') {
+        // NOTE(pahaz): We want to replace the current error if we determine that we received a GQLError.
+        // For example, when implementing the logic for GQLCustomSchema, if a call to User.create() occurs during the execution of a query
+        // which triggers a GQLError validation error, we want to pass this error up in the response so that it is visible to the user.
+        if (originalErrorCName === 'GQLError') {
+            const code = originalError?.extensions?.code
+            const type = originalError?.extensions?.type
+            const parentErrors = originalError?.errors
+            // We have an exact one parent gqlError.errors, and it is GQLError({ code: INTERNAL_ERROR, type: SUB_GQL_ERROR })
+            if (code === 'INTERNAL_ERROR' && type === 'SUB_GQL_ERROR' && isArray(parentErrors) && parentErrors.length === 1) {
```

Comment:
What about `else` here?

Reply:
We don't need it here. It's just a special case for internal errors.